src/s1reader/s1_burst_slc.py

```diff
@@ -579,6 +585,332 @@ def az_carrier_components(self, offset, position):
+                    + (arr_polynomial[:,1][..., np.newaxis] 
+                       * np.ones(ncol)[np.newaxis, ...]) * term_tau
+                    + (arr_polynomial[:,2][..., np.newaxis]
+                       * np.ones(ncol)[np.newaxis, ...]) * term_tau**2)
+
+        return eval_out
+
+    def _llh_to_ecef(self, lat, lon, hgt, ellipsoid, unit_degree=True):
+        '''
+        Calculate cartesian coordinates in ECEF from
+        latitude, longitude, and altitude
+        
+        Parameters:
+        -----------
+        lat: np.ndarray
+            latitude as numpy array
+        lon: np.ndarray
+            longitude as numpy array
+        hgt: np.ndarray
+            height as numpy array
+        ellipsoid: isce3.core.Ellipsoid
+            Definition of Ellipsoid
+        unit_degree: bool
+            True if the units of lat and lon are degrees.
+            False if the units are radian.
+        
+        Return:
+        _: tuple(np.ndarray)
+            x, y, and z as a tuple of np.ndarray
+        
+        '''
+
+        if unit_degree:
+            rad_lat = lat * (np.pi / 180.0)
+            rad_lon = lon * (np.pi / 180.0)
+        else:
+            rad_lat = lat
+            rad_lon = lon
+
+        v = ellipsoid.a / np.sqrt(1 - ellipsoid.e2 * np.sin(rad_lat) * np.sin(rad_lat))
+
+        x = (v + hgt) * np.cos(rad_lat) * np.cos(rad_lon)
+        y = (v + hgt) * np.cos(rad_lat) * np.sin(rad_lon)
+        z = (v * (1 - ellipsoid.e2) + hgt) * np.sin(rad_lat)
+
+        return (x, y, z)
+
+    def az_fm_rate_mismatch_mitigation(self, path_dem: str, path_scratch: str,
+            threshold_rdr2geo = 1e-8,
+            numiter_rdr2geo = 25,
+            custom_radargrid = None):
+        '''
+        Calculate azimuth FM rate mismatch mitigation
+        Based on ETAD-DLR-DD-0008, Algorithm Technical Baseline Document.
+        Available: https://sentinels.copernicus.eu/documents/247904/4629150/
+        Sentinel-1-ETAD-Algorithm-Technical-Baseline-Document.pdf
+
+        Parameters:
+        -----------
+        path_dem: str
+            Path to the DEM to calculate the actual azimuth FM rate
+        path_scratch: str
+            Path to the scratch directory to store intermediate data
+        threshold_rdr2geo: int
+            Threshold of the iteration for rdr2geo
+        numiter_rdr2geo: int
+            Maximum number of iteration for rdr2geo
+        custom_radargrid_correction: isce3.product.RadarGridParameters
+            ISCE3 radar grid to define the correction grid
+        
+        Return:
+        -------
+        _: isce3.core.LUT2d
+            azimuth FM rate mismatch rate in radar grid in seconds.
+
+        '''
+        
+        os.makedirs(path_scratch, exist_ok=True)
+        
+        if custom_radargrid is None:
+            radargrid_correction = self.as_isce3_radargrid()
+        else:
+            radargrid_correction = custom_radargrid
```

Comment:
Is this taking into account the possibility of generating the correction on a coarser radar grid?

Reply:
Yes. Also we can compare our calculation results with ETAD product by defining `custom_radargrid` in accordance with their product's correction grid.

Reply:
@vbrancat Just to note, I've revised the function `az_fm_rate_mismatch_mitigation` to optionally take in range / azimuth steps, so that its interface looks more similar to `bistatic_delay` and `geometrical_and_steering_doppler`. Although, the azimuth FM rate calculation still needs DEM to calculate.